Port RemoteScrollingUIState to the new IPC serialization format

cdumez opened the pull request:
#### 94a4dd856588b80b2f7ca42d2d9facb62b67fa9d
<pre>
Port RemoteScrollingUIState to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=264523">https://bugs.webkit.org/show_bug.cgi?id=264523</a>

Reviewed by Alex Christensen.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.cpp:
(WebKit::RemoteScrollingUIState::RemoteScrollingUIState):
(WebKit::emptyNodesHashSet):
(WebKit::RemoteScrollingUIState::nodesWithActiveScrollSnapIfChanged const):
(WebKit::RemoteScrollingUIState::nodesWithActiveUserScrollsIfChanged const):
(WebKit::RemoteScrollingUIState::nodesWithActiveRubberbandIfChanged const):
(WebKit::RemoteScrollingUIState::encode const): Deleted.
(WebKit::RemoteScrollingUIState::decode): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/270506@main">https://commits.webkit.org/270506@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61dbeecb91fe05d863271ed54e05f12281de1ec0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4283 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26961 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27778 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23518 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1718 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22128 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28358 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29159 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23431 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23449 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27018 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2841 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1077 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4223 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3291 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3281 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3157 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.serialization.in:
```diff
@@ -0,0 +1,35 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+class WebKit::RemoteScrollingUIState {
+    OptionSet<WebKit::RemoteScrollingUIStateChanges> changes();
```
This is what OptionalTupleBits/OptionalTupleBit are for.  Then we don't need emptyNodesHashSet().  We can just change the branch in the accessors to ASSERT(m_changes.contains(Changes::ScrollSnapNodes)) etc.

achristensen07 commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.h:
```diff
@@ -50,8 +50,8 @@ class RemoteScrollingUIState {
 
-    void encode(IPC::Encoder&) const;
-    static std::optional<RemoteScrollingUIState> decode(IPC::Decoder&);
+    RemoteScrollingUIState() = default;
```
Do we still need this?

cdumez replied:
Yes, since I'm adding a constructor that takes parameters, I need to explicitly define the default constructor to avoid breaking the build.

cdumez committed:
Port RemoteScrollingUIState to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=264523

Reviewed by Alex Christensen.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.cpp:
(WebKit::RemoteScrollingUIState::RemoteScrollingUIState):
(WebKit::emptyNodesHashSet):
(WebKit::RemoteScrollingUIState::nodesWithActiveScrollSnapIfChanged const):
(WebKit::RemoteScrollingUIState::nodesWithActiveUserScrollsIfChanged const):
(WebKit::RemoteScrollingUIState::nodesWithActiveRubberbandIfChanged const):
(WebKit::RemoteScrollingUIState::encode const): Deleted.
(WebKit::RemoteScrollingUIState::decode): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingUIState.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/270506@main
